File: 32.最长有效括号.py
#
# @lc app=leetcode.cn id=32 lang=python3
#
# [32] 最长有效括号
#

# @lc code=start
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        # 动态规划
        # ans = 0
        # size = len(s)
        # dp = [0] * size

        # for i in range(1, size):
        #     if s[i] == ')':
        #         if s[i-1] == '(':
        #             dp[i] = (dp[i-2] if i >= 2 else 0) + 2
        #         elif i - dp[i-1] > 0 and s[i-dp[i-1]-1] == '(':
        #             dp[i] = dp[i-1] + (dp[i-dp[i-1]-2] if i-dp[i-1]>2 else 0) + 2
        #         ans = max(ans, dp[i])

        # return ans
        
        # 栈
        # ans = 0
        # size = len(s)
        # stack = list()
        # stack.append(-1)
        # for i in range(size):
        #     if s[i] == '(':
        #         stack.append(i)
        #     else:
        #         stack.pop()
        #         if not stack:
        #             stack.append(i)
        #         else:
        #             ans = max(ans, i - stack[-1])

        # return ans

        # 双向遍历
        left, right, ans = 0, 0, 0
        size = len(s)
        for i in range(size):
            if s[i] == '(':
                left += 1
            else:
                right += 1
            if left == right:
                ans = max(ans, 2 * right)
            elif right > left:
                left, right = 0, 0
        left, right = 0, 0
        for i in range(size-1, -1, -1):
            if s[i] == '(':
                left += 1
            else:
                right += 1
            if left == right:
                ans = max(ans, 2 * left)
            elif left > right:
                left, right = 0, 0

        return ans


# @lc code=end

File: 32.最长有效括号.java
/*
 * @lc app=leetcode.cn id=32 lang=java
 *
 * [32] 最长有效括号
 */

// @lc code=start
class Solution {
    public int longestValidParentheses(String s) {
        // left right
        int left = 0, right = 0, ans = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                left++;
            } else {
                right++;
            }
            if (left == right) {
                ans = Math.max(ans, 2 * right);
            } else if (right >= left) {
                left = right = 0;
            }
        }

        left = right = 0;
        for (int i = s.length()-1; i>= 0; i--) {
            if (s.charAt(i) == '(') {
                left++;
            } else {
                right++;
            }
            if (left == right) {
                ans = Math.max(ans, 2 * left);
            } else if (left >= right) {
                left = right = 0;
            }
        }
        return ans;

        // stack
        // int ans = 0;
        // Stack<Integer> stack = new Stack<>();
        // stack.push(-1);
        // for (int i = 0; i < s.length(); i++) {
        //     if (s.charAt(i) == '(') {
        //         stack.push(i);
        //     } else {
        //         stack.pop();
        //         if (stack.empty()) stack.push(i);
        //         else ans = Math.max(ans, i-stack.peek());
        //     }
        // }
        // return ans;

        // DP
        // int ans = 0;
        // int[] dp = new int[s.length()];
        // for (int i = 1; i < s.length(); i++) {
        //     if (s.charAt(i) == ')') {
        //         if (s.charAt(i-1) == '(') {
        //             dp[i] = (i >= 2 ? dp[i-2] : 0) + 2;
        //         } else if (i-dp[i-1] > 0 && s.charAt(i-dp[i-1]-1) == '(') {
        //             dp[i] = dp[i-1] + ((i-dp[i-1]) >= 2 ? dp[i-dp[i-1]-2] : 0) + 2;
        //         }
        //         ans = Math.max(ans, dp[i]);
        //     }
        // }
        // return ans;
    }
}
// @lc code=end

